### ISSUE TITLE
[Bug]: Does the function 'isFormatSupported' work correctly?

### ISSUE THREAD

**Issue by gizomo** [TheWidlarzGroup/react-native-video#3311]
### What happened?

The function `isFormatSupported` for ExoPlayer returns true if it has failed to get decoder info or mimeType. But it returns false if the format has no width, height or frameRate although the format is supported by the codec (we can check `codecInfo.isFormatSupported(format)` and it returns true).

I have an m3u8 playlist:

```
#EXTM3U
#EXT-X-VERSION:3

#EXT-X-STREAM-INF:BANDWIDTH=1728000
633812eb_83fc_4b3b_a2a3_64371780bd91.m3u8?bandwidth=1728&e=1698226963&max=0&min=0&token=84b6de50070acde8d9eda2619c5658bc

#EXT-X-STREAM-INF:BANDWIDTH=928000
633812eb_83fc_4b3b_a2a3_64371780bd91.m3u8?bandwidth=928&e=1698226963&max=0&min=0&token=84b6de50070acde8d9eda2619c5658bc

#EXT-X-STREAM-INF:BANDWIDTH=528000
633812eb_83fc_4b3b_a2a3_64371780bd91.m3u8?bandwidth=528&e=1698226963&max=0&min=0&token=84b6de50070acde8d9eda2619c5658bc
```

There are three video tracks. But all of them are skipped because `codecInfo.isVideoSizeAndRateSupportedV21(width, height, frameRate)` returns false, as width, height and frameRate are set to `Format.NO_VALUE`. If we check `codecInfo.isFormatSupported(format)` it returns true.

### Steps to reproduce

1. Use an HLS (.m3u8) master playlist whose `#EXT-X-STREAM-INF` variants declare only `BANDWIDTH` (no width, height or frame rate), so the resulting `Format` has width/height/frameRate set to `Format.NO_VALUE`.
2. Play the source on Android with ExoPlayer.
3. Observe that all video tracks are skipped because `isFormatSupported` returns false via `codecInfo.isVideoSizeAndRateSupportedV21(width, height, frameRate)`.

### Platforms

Android

### Media / source type

HLS (.m3u8)

**Comment by freeboub** [comment]
> Function 'isFormatSupported' for exoplayer returns true if it has failed to get decoder info or mimeType.

I agree it look like an issue. but I think it was made only to bypass possible MediaCodec issue. 
But It would need a specific error log !

I think an acceptable workaround would be to consider video track supported if width & height == 0. (directly tested in isFormatSupported).

The best would be to fix playlist server side !

**Comment by gizomo** [comment]
> The best would be to fix playlist server side !

Sometimes its impossible. For instance when we use third party content. I'm afraid that the vast majority of playlists are designed as simply as I gave the example above

**Comment by gizomo** [comment]
Here is a simplified version of 'isFormatSupported' function (in my project I’m also checking for softcodecs support).
```
private boolean isFormatSupported(Format format) {
    if (format.roleFlags == C.ROLE_FLAG_TRICK_PLAY) {
        return false;
    }

    int width = format.width == Format.NO_VALUE ? 0 : format.width;
    int height = format.height == Format.NO_VALUE ? 0 : format.height;
    float frameRate = format.frameRate == Format.NO_VALUE ? 0 : format.frameRate;
    String mimeType = format.sampleMimeType;

    if (mimeType == null) {
        return true;
    }

    boolean isSupported;

    try {
        List<MediaCodecInfo> codecsInfoList = MediaCodecUtil.getDecoderInfos(mimeType, false, false);

        for (int i = 0; i < codecsInfoList.size(); ++i) {
            MediaCodecInfo codecInfo = codecsInfoList.get(i);

            if (codecInfo.isVideoSizeAndRateSupportedV21(width, height, frameRate) || codecInfo.isFormatSupported(format)) {
                isSupported = true;
            }
        }
    } catch (Exception e) {
        // Failed to get decoder info - assume it is supported
        isSupported = true;
    }

    return isSupported;
}
```

**Comment by freeboub** [comment]
@gizomo I wonder why you consider ROLE_FLAG_TRICK_PLAY track as not supported.

It would be more interesting to report role flag to application ? (I did it on my work fork) 
```
    if (format.roleFlags == C.ROLE_FLAG_TRICK_PLAY) {
        return false;
    }
```

A small improvement can be to break once a codec is defined as supported:

```
 for (int i = 0; i < codecsInfoList.size(); ++i) {
            MediaCodecInfo codecInfo = codecsInfoList.get(i);
            if (codecInfo.isVideoSizeAndRateSupportedV21(width, height, frameRate) || codecInfo.isFormatSupported(format)) {
                isSupported = true;
+              break;
            }
        }
```

**Comment by TWG-AI** [comment]
Thanks for the detailed analysis and the simplified code snippet! This has been logged for the team.

The root issue is clear: `isVideoSizeAndRateSupportedV21` is being called with `Format.NO_VALUE` for width/height/frameRate instead of treating those as unconstrained. The fix direction would be to skip the size/rate check when any of those values is `Format.NO_VALUE` and fall back to `codecInfo.isFormatSupported(format)` directly — exactly what your simplified version shows.

Unfortunately there's no simple workaround for third-party manifests where you can't add `RESOLUTION` to the `#EXT-X-STREAM-INF` entries.

> I'm an automated triage assistant — please double-check the details here. A human maintainer will look into a fix.
